README.md
# PMD-Roleplay Site Beta Userscript File
A userscript containing beta features of the PMD-roleplay.forumotion.org site.
## Installation
You must have an extension on your webbrowser that supports userscripts. This was developed and tested on Tampermonkey, which is a fork of greasemonkey for Chrome.

Usually, if you have these extensions installed. You may install the script simply by visiting the script source. This will open up the extension and display the scripts meta information and give you an option to install it. Please use [this link](https://github.com/coaster3000/pmd-roleplay-beta/raw/releases/dist/pmd-roleplay-beta.user.js) to open the latest distributed source.


CHANGELOG.MD
# Changelogs
List of changes between versions of the script.

## 0.0.6
- Made userscript entirely modular.
  - SiteAjax converted into module.
  - ToggleAnimations converted into module.
    - Module disabled by default. To enable use ```AnimationModule.load();```

## 0.0.5
- Script Meta Data Changes
  - Added supportURL metadata.
  - Added License metadata.
  - Switched distribution to openuserjs.

## 0.0.4
- Improved link detection for AJAX features.

## 0.0.3
**Critical**  Fixed URL mismatch

## 0.0.2
Ajaxified the site.

## 0.0.1
Initial releases


Gruntfile.js
module.exports = function(grunt) {
	var defaultTasks = ['concat'/*, 'uglify'*/]; // used for watch as well

	var files = [ 'src/UserScriptHeader.js', /*'src/Header.js',*/
				'src/Main/*.js'/*,
				'src/Footer.js'*/ ];

	var opkg = grunt.file.readJSON('package.json');
	grunt.initConfig({
		pkg: opkg,

		concat: {
			options: {
				process: function(src, filepath) {
					return (filepath != "src/UserScriptHeader.js"?'\n// Source: ' + filepath + '\n\n' : "") + src.replace(/@VERSION@/g, opkg.version);
				}
			},
			all: {
				src: files,
				dest: 'dist/<%= pkg.name %>.user.js',
			},
			meta: {
				src: ['src/UserScriptHeader.js'],
				dest: 'dist/<%= pkg.name %>.meta.js'
			}
		},

		/*uglify: {
			all: {
				files: {
					'pmd-roleplay-extended-chat.min.user.js': ['pmd-roleplay-extended-chat.user.js']
				}
			}
		},*/

		watch: {
			all: {
				files: files,
				tasks: defaultTasks,
				options: {
					debounceDelay: 250
				}
			}
		}
	});

	grunt.loadNpmTasks('grunt-contrib-concat');
	//grunt.loadNpmTasks('grunt-contrib-uglify');
	grunt.loadNpmTasks('grunt-contrib-watch');

	grunt.registerTask('default', defaultTasks);
};


src/Main/SiteAjax.js
var ajaxer = {page: null, title: null, href: null};

function ajaxComplete() {
  $("#loader").remove();
  window.history.replaceState({page: ajaxer.page}, ajaxer.title, ajaxer.href);
  document.title=this.textContent;
  initAjax();
}

function firstInitAjax() {
  $("a.mainmenu,a.nav,a.forumlink").click(onAjaxClick);
}

function onAjaxClick(e) {
  //Display Loader ICON
  var loaderICON = "http://i.imgur.com/6ZwRMCn.gif";
  $("body").append(`<div id="loader" style="display: block; width:100%; height: 100%; background-color: rgba(0,0,0,0.3); z-index: 150; position: fixed; top:0; left: 0; padding: 0; margin: 0; text-align: center;"><img src="${loaderICON}" style="margin: auto; position: relative; top: 50%"/></div>`);

  ajaxer.page="forum";
  ajaxer.title=this.textContent;
  ajaxer.href=this.href;
  $("#page-body").load(this.href + " #page-body > div", ajaxComplete);
  return false;
}
function initAjax(){

  if ($("#chatbox_top").length > 0) { //Chatbox re-init
    $("#chatbox_top").html(`<iframe src="/chatbox/index.forum?page=front&amp;" id="frame_chatbox" scrolling="no" width="100%" height="100%" marginwidth="0" marginheight="0" frameborder="0"></iframe>`);
  }

  $("a.nav,a.forumlink").click(onAjaxClick);
}

firstInitAjax();


src/UserScriptHeader.js
// ==UserScript==
// @name	pmd-rp-beta
// @namespace	coaster3000@pmd-roleplay.forumotion.com
// @author coaster3000
// @description	A userscript that provides beta features for the site pmd-roleplay.forumotion.org. All beta features are eventually added to site once enough feedback is provided on site from userscript testing.
// @include	http://pmd-roleplay.forumotion.org/*
// @include	https://pmd-roleplay.forumotion.org/*
// @match	https://pmd-roleplay.forumotion.org/*
// @match	http://pmd-roleplay.forumotion.org/*
// @version	@VERSION@
// @license MIT; https://github.com/coaster3000/pmd-roleplay-beta/blob/master/LICENSE.md
// @supportURL https://github.com/coaster3000/pmd-roleplay-beta/issues
// @updateURL https://openuserjs.org/meta/coaster3000/pmd-rp-beta.meta.js
// @downloadURL https://openuserjs.org/install/coaster3000/pmd-rp-beta.user.js
// @grant	none
// ==/UserScript==


src/Main/ToggleAnimations.js
const PMD_ANIMATE_COOKIE = "PMD-Animate";
const ANIM_DISABLE_TXT = "Disable Animation";
const ANIM_ENABLE_TXT = "Enable Animation";
function setAnimations(value) {
	var state = (my_getcookie(PMD_ANIMATE_COOKIE) > 0);
	if (value != state) {
		my_setcookie(PMD_ANIMATE_COOKIE, (value?1:0), true);
		updateAnimations();
	}
}

function updateAnimations() {
	var state = ((my_getcookie(PMD_ANIMATE_COOKIE) || 0) > 0);
	if (state)
		$(".animation").addClass("animated");
	else
		$(".animation").removeClass("animated");

	$("#ToggleAnimationsBtn").val(state?ANIM_DISABLE_TXT:ANIM_ENABLE_TXT);
}

function toggleAnimations() {
	var btn = $("#ToggleAnimationsBtn");
	if (btn.val()==ANIM_ENABLE_TXT) {
		setAnimations(true);
	} else {
		setAnimations(false);
	}

}

function createAnimationsButton() {
	$("body").append('<input type="button" name="ToggleAnimations" id="ToggleAnimationsBtn" value="Toggle Animations"/>');
	$("#ToggleAnimationsBtn").on("click", toggleAnimations);
}

createAnimationsButton();
updateAnimations();


src/Main/01_ToggleAnimations.js
var AnimationModule = (function(setcookie,getcookie,$) {
	'use strict';

	var PMD_ANIMATE_COOKIE = "PMD-Animate";
	var ANIM_DISABLE_TXT = "Disable Animation";
	var ANIM_ENABLE_TXT = "Enable Animation";

	function setAnimations(value) {
		var state = (getcookie(PMD_ANIMATE_COOKIE) > 0);
		if (value != state) {
			setcookie(PMD_ANIMATE_COOKIE, (value?1:0), true);
			updateAnimations();
		}
	}

	function updateAnimations() {
		var state = ((getcookie(PMD_ANIMATE_COOKIE) || 0) > 0);
		if (state)
		$(".animation").addClass("animated");
		else
		$(".animation").removeClass("animated");

		$("#ToggleAnimationsBtn").val(state?ANIM_DISABLE_TXT:ANIM_ENABLE_TXT);
	}

	function toggleAnimations() {
		var btn = $("#ToggleAnimationsBtn");
		if (btn.val()==ANIM_ENABLE_TXT) {
			setAnimations(true);
		} else {
			setAnimations(false);
		}

	}

	function createAnimationsButton() {
		$("body").append('<input type="button" name="ToggleAnimations" id="ToggleAnimationsBtn" value="Toggle Animations"/>');
		$("#ToggleAnimationsBtn").on("click", toggleAnimations);
	}



	function load() {
		createAnimationsButton();
		updateAnimations();
	}

	return {load: load};
})(my_setcookie, my_getcookie,jQuery); // jshint ignore:line

// AnimationModule.load();


src/Main/02_SiteAjax.js
var SiteAjax = (function($,w){
  'use strict';
  var ajaxer = {page: null, title: null, href: null};

  function checkForProperties(obj, props) {
    if (obj == null || typeof obj == "undefined") return false;
    if (typeof props == "string") return (obj.hasOwnProperty(props) || checkForProperties(Object.getPrototypeOf(obj), props));
    else if (!Array.isArray(props)) throw "Illegal Argument Error: props parameter is not valid, must be a string or an Array of strings.";
    for (var i in props) {
      if (!checkForProperties(obj, props[i]) && !checkForProperties(Object.getPrototypeOf(obj), props[i])) return false;
    }
    return true;
  }

  function displayLoader() {
    var loaderICON = "http://i.imgur.com/6ZwRMCn.gif";
    $("body").append(`<div id="loader" style="display: block; width:100%; height: 100%; background-color: rgba(0,0,0,0.3); z-index: 150; position: fixed; top:0; left: 0; padding: 0; margin: 0; text-align: center;"><img src="${loaderICON}" style="margin: auto; position: relative; top: 50%"/></div>`);
  }

  function hideLoader() {
    $("#loader").remove();
  }

  function loadComplete() {
    hideLoader();
    w.history.replaceState({page: ajaxer.page}, ajaxer.title, ajaxer.href);
    document.title=ajaxer.title;
    initAjax();
  }

  function loadPage(a) { // a should be an html "a" element
    if (!checkForProperties(a, ["href","textContent"])) throw "Invalid Parameter: a is not a valid object.";

    displayLoader();

    ajaxer = {page: "forum", title: a.textContent, href: a.href};

    $("#page-body").load(a.href + " #page-body > div", loadComplete); // Actual load operation
  }


  function clickHandler() {
    if (typeof w.history == "undefined" || typeof w.history.replaceState == "undefined") return true; // AJAX loading not supported fully. Just reload the page instead.

    loadPage(this); // jshint ignore:line
    return false; // Cancel normal navigation.
  }

  function initAjax(){
    if ($("#chatbox_top").length > 0) { // Reinitiate chatbox if found.
      $("#chatbox_top").html(`<iframe src="/chatbox/index.forum?page=front&amp;" id="frame_chatbox" scrolling="no" width="100%" height="100%" marginwidth="0" marginheight="0" frameborder="0"></iframe>`);
    }

    $("a").each(function(i,e) {
      if (e.host=="pmd-roleplay.forumotion.org" && /https?:\/\/pmd-roleplay\.forumotion\.org\/[ft].+/.test(e.href)) {

        $(e).unbind("click.pmdAjax").bind("click.pmdAjax", clickHandler);
      }
    });
  }

  function init() {
    initAjax();
  }

  return {init: init};
})(jQuery, window);
SiteAjax.init();
